runtime/src/secondary_index.rs

```diff
@@ -99,12 +100,138 @@ impl SecondaryIndexEntry for RwLockSecondaryIndexEntry {
+        &self.map
+    }
+    fn sort_slice_by_value(&mut self, slice_key: &L1) {
+        // Obtain a slice of mutable refrences to all elements with the same key
+        let mut sub_slice = self
+            .map
+            .split_mut(|(k, _)| *k != *slice_key)
+            .flatten()
+            .collect_vec();
+        // Sort them
+        let mut sorting_buffer = sub_slice.iter().map(|x| *(*x)).collect_vec();
+        sorting_buffer.sort_unstable_by_key(|(_, v)| *v);
+        // Copy back into the list
+        for i in 0..sub_slice.len() {
+            *(sub_slice[i]) = sorting_buffer[i];
+        }
+    }
+    fn update_map(&mut self, key: &L1, value: &L2) {
+        // Check if the value already exists.
+        let existing_value_position = self.map.iter().position(|(_, y)| *y == *value);
+        // Remove it if it does.
+        // Note: Removal maintains sorted order, updating would require a re-sort.
+        // Thus, since we have to search to find the new position anyways,
+        // just throw it away and re-insert as if its a new element.
+        if let Some(position) = existing_value_position {
+            self.map.remove(position);
+        }
+        // If its a new value...
+        else {
+            // Check if the list is full, and if the key is less than the smallest element, if so exit early.
+            if self.map.len() >= self.capacity && self.map[0].0 > *key {
+                return;
+            }
+        };
+        // Is this an excluded key?
+        if self.excluded_keys.contains(key) {
+            return;
+        }
+        // Check if there are more elements in the list with the same key
+        let existing_key_position = self.map.iter().position(|(x, _)| *x == *key);
+        // Find where the new entry goes and insert it.
+        let key_position = match self.map.binary_search_by_key(key, |&(k, _)| k) {
+            Ok(found_position) => found_position,
+            Err(woudbe_position) => woudbe_position,
+        };
+        self.map.insert(key_position, (*key, *value));
+        // If there was indeed more elements with the same key sort them by value
+        if existing_key_position.is_some() {
+            self.sort_slice_by_value(key);
+        }
+        // Prune list of too big
+        while self.map.len() > self.capacity {
+            self.map.remove(0);
+        }
+    }
+}
+
+pub type SecondaryIndexKeySizes = RwLock<HierarchicalOrderedMap<usize, Pubkey>>;
+
+pub trait SecondaryIndexLargestKeys: Debug {
+    const MAX_NUM_RETURNED_LARGEST_KEYS: usize;
+    const MAX_NUM_LARGEST_KEYS: usize;
+    fn get_largest_index_keys(&self, max_entries: usize) -> Vec<(usize, Pubkey)>;
+    fn update_largest_index_keys(&self, key_size: &usize, pubkey: &Pubkey);
+    fn default() -> Self;
+}
+
+impl SecondaryIndexLargestKeys for SecondaryIndexKeySizes {
+    const MAX_NUM_RETURNED_LARGEST_KEYS: usize = 10;
+    const MAX_NUM_LARGEST_KEYS: usize = 100;
```

Comment:
Maybe just move these to module-level constants, so they aren't so difficult to import in your test, and don't require the other bits to be pub. They will probably need renaming, though.